docs/architecture/adr-103-proto-versioning.md

```diff
@@ -0,0 +1,140 @@
+releases.
+
+We hypothesize that, if we had to version our Protobuf definitions according to
+[Buf's style guide][buf-style], we could simplify integration efforts with
+CometBFT in the long run. Adopting such a practice would also introduce a new
+level of rigour into the way CometBFT Protobuf definitions are managed, allowing
+for better expectation management with integrators. It would also allow the core
+team to make better use of Protobuf-related tooling (such as Buf) in enforcing
+standards, conventions and versioning practices in an automated way.
+
+Some context is captured in [\#95] and the issues/PRs to which it links.
+
+## Alternative Approaches
+
+1. The main alternative here is to do nothing and keep our existing Protobuf
+   definition approach, which suffers from the issues captured in the
+   [Context](#context) section.
+2. Version our Protobuf definitions, as per this ADR. The primary drawback of
+   introducing versioning is that it represents a substantial breaking change.
+   Within this approach, there are two alternative policies that @mzabaluev has
+   proposed:
+   1. **Conservative**, recommended by Buf: Do not create new versions when
+      making non-breaking changes to Protobuf message definitions, such as
+      adding a new field. The upside to this approach is that non-breaking
+      changes can be made in non-breaking releases of CometBFT, but the downside
+      is that some programming languages, such as Rust, cannot generate code in
+      a non-breaking way from these non-breaking changes.
+   2. **Sensitive**: Create a new version of a Protobuf message definition any
+      time anything changes, including adding a new field. The upside to this
+      approach is that generated code in languages like Rust is additive and
+      non-breaking, but the downside is that non-breaking Protobuf changes can
+      only be released in new, breaking versions of CometBFT.
+
+## Decision
+
+From team discussion, it was decided to adopt versioning, using the **sensitive
+versioning** policy.
+
+## Detailed Design
+
+### Implementation
+
+In order to implement Protobuf definition versioning, it is recommended to bring
+all of the Protobuf definitions for all major CometBFT releases thus far (v0.34,
+v0.37 and v0.38) into `main`, implementing versioning for each major version's
+definitions.
+
+For example, the v0.34 Protobuf definitions would be brought to `main` as `v1`
+(i.e. `proto/tendermint/types/types.proto` would become
+`proto/tendermint/types/v1/types.proto`, and the package would change from
+`tendermint.types` to `tendermint.types.v1`).
+
+Then, the v0.37 Protobuf definitions would be brought to `main`, and only where
+changes were made according to our **sensitive** versioning policy would we
+create new types (e.g. `proto/tendermint/types/params.proto` from v0.37 would
+become `proto/tendermint/types/v2/params.proto` and the package would change
+from `tendermint.types` to `tendermint.types.v2`).
+
```

Comment:
Could you add a paragraph explaining that, when a new version of a protobuf package is introduced, all changes w.r.t. the previous version MUST be documented:
* How to interpret a message using the old version when the new version is expected
* And vice versa
* The interpretation for either of the above could simply be "REJECT", meaning that this is not expected to happen in production.
(let me know if you'd rather have me add that paragraph)

Reply:
This information should rather exist in the README in the `proto` folder. Otherwise people may miss it, given that it'd be buried in an ADR.